test/pipelinerun_test.go

```diff
@@ -160,14 +161,36 @@ func TestPipelineRun(t *testing.T) {
 				if !r.Status.GetCondition(duckv1alpha1.ConditionSucceeded).IsTrue() {
 					t.Fatalf("Expected TaskRun %s to have succeeded but Status is %v", taskRunName, r.Status)
 				}
-				for name, key := range map[string]string{
-					pipelineRunName: pipeline.PipelineRunLabelKey,
-					pipelineName:    pipeline.PipelineLabelKey,
-				} {
-					expectedName := getName(name, i)
-					lbl := pipeline.GroupName + key
-					if val := r.Labels[lbl]; val != expectedName {
-						t.Errorf("Expected label %s=%s but got %s", lbl, expectedName, val)
+
+				// Check label propagation to Tasks
+				labels := make(map[string]string, len(pr.ObjectMeta.Labels)+3)
+				for key, val := range pr.ObjectMeta.Labels {
+					labels[key] = val
+				}
+				labels[pipeline.GroupName+pipeline.PipelineLabelKey] = getName(pipelineName, i)
+				labels[pipeline.GroupName+pipeline.PipelineRunLabelKey] = getName(pipelineRunName, i)
+				for key, pipelineRunVal := range labels {
+					if taskRunVal := r.ObjectMeta.Labels[key]; taskRunVal != pipelineRunVal {
+						t.Errorf("Expected label %s=%s but got %s", key, pipelineRunVal, taskRunVal)
+					}
+				}
+
+				// Check label propagation to Pods
+				taskRunLabelKey := pipeline.GroupName + pipeline.TaskRunLabelKey
+				labels[taskRunLabelKey] = taskRunName
+				pods, err := c.KubeClient.Kube.CoreV1().Pods(namespace).List(metav1.ListOptions{
+					LabelSelector: taskRunLabelKey + " = " + taskRunName,
+				})
+				if err != nil {
+					t.Fatalf("Couldn't get expected Pod for %s: %s", taskRunName, err)
+				}
+				if numPods := len(pods.Items); numPods != 1 {
+					t.Fatalf("Expected 1 Pod for %s, but got %d Pods", taskRunName, numPods)
+				}
+				pod := pods.Items[0]
+				for key, taskRunVal := range labels {
+					if podVal := pod.ObjectMeta.Labels[key]; podVal != taskRunVal {
+						t.Errorf("Expected label %s=%s but got %s", key, taskRunVal, podVal)
```

Comment:
total nitpick: could these be in a separate function called by the test? for the same of readability (test is getting long!)

Reply:
Yeah good point, I'll separate it out!

Reply:
Did some refactoring in 37ed488, let me know if that seems good or if you'd like to see a totally separate `assertLabels` function with all the label-related stuff to simplify further.